File: 2b/sixLittle.rb
class SixLittle
  # Constructor that initializes the instance variables to empty arrays
  def initialize
    @words = []
    @hints = []
    @tokens = []
  end

  def ask
    6.times do
      puts "Enter a word (at least 4 characters):"
      word = gets.chomp
      # Keep asking for a word UNTIL the word is at least 4 characters
      until word.length >= 4
        puts "Word too short. Enter a word (at least 4 letters):"
        word = gets.chomp
      end
      # Word was at least 4 characters, so ask for a hint for that word now
      puts "Enter a hint for this word:"
      hint = gets.chomp

      # Add the word and hint to the appropriate arrays (instance variables)
      @words << word.upcase
      @hints << hint.capitalize
    end
  end

  def prepare
    # for EACH word in the words array, do...
    @words.each do |word|
      # Split the word in half (rounding up if odd number of characters), always rounding up to be consistent
      mid = (word.length / 2.0).ceil
      # Uppercase the first half and the second half of the word and add them to the tokens array
      @tokens << word[0...mid].upcase
      @tokens << word[mid..-1].upcase
    end
    # Shuffle the tokens array
    @tokens = @tokens.shuffle
  end

  def display
    puts "Six Little Words (Ruby)"
    # Slice the tokens array into groups of 4, then for each group/slice, join the 4 tokens with a space and join the slices with a newline
    puts "Tokens:\n " + @tokens.each_slice(4).map { |slice| slice.join(' ') }.join("\n")

    puts "\nHints:"
    # For each hint in the hints array, print the hint
    @hints.each { |hint| puts hint }

    puts "\nAnswer Key:"
    # Slice the words array into groups of 3, then for each group/slice, join the 3 words with a space and join the slices with a newline
    puts @words.each_slice(3).map { |slice| slice.join(' ') }.join("\n")
  end

  def run
    # Call the methods, weird parenthesis aren't needed even with no arguments...
    ask
    prepare
    display
  end
end

# Create an new SixLittle object and call the run method of that object to start the program
six_little = SixLittle.new
six_little.run

File: Scala/5a/sixLittle.scala
import scala.collection.mutable.ArrayBuffer
import scala.io.StdIn.readLine
import scala.util.Random

class SixLittle {
    // mutable arrays
    val words = ArrayBuffer[String]()
    val hints = ArrayBuffer[String]()
    val tokens = ArrayBuffer[String]()

    // Ask method
    def ask(): Unit = {
    // Asking 6 times
        for (_ <- 0 until 6) {
            var word = ""
            // while word is less than 4 characters... ask for a word again
            while (word.length < 4) {
                print("Enter a word: ")
                word = readLine()
                if (word.length < 4) {
                    println("Word must be at least 4 characters long.")
                }
            }
            words += word

            // Ask for a hint
            print("Enter a hint: ")
            val hint = readLine()
            hints += hint
        }
    }


    // prepare method
    def prepare(): Unit = {
        words.foreach { word =>
            // Split word in half and convert to upper case
            val half = (word.length + 1) / 2
            // add first half to tokens
            tokens += word.substring(0, half).toUpperCase()
            // Add second half to tokens
            tokens += word.substring(half).toUpperCase()
        }

        // Shuffle tokens array
        Random.shuffle(tokens)
    }

    // display method
    def display(): Unit = {
        println("Six Little Words (Scala)\n")
        println("Partial Words/Tokens:")
        // Group tokens by 4
        tokens.grouped(4).foreach { group =>
            // Print each group separated by a space
            println(group.mkString(" "))
        }

        println("\nHints:")
        // for each hint, print it
        hints.foreach { hint =>
            println(hint)
        }

        println("\nAnswer Key:")
        // Group words by 3
        words.grouped(3).foreach { group =>
            // Print each group separated by a space
            println(group.mkString(" "))
        }
    }

    // Run method
    def run(): Unit = {
        ask()
        prepare()
        display()
    }
}

// Main object
object SixLittle {
  def main(args: Array[String]): Unit = {
    val game = new SixLittle()
    game.run()
  }
}
